localboot/main.go

```diff
@@ -2,56 +2,43 @@ package main
 
 import (
 	"flag"
+	"fmt"
 	"log"
+	"os"
 	"path"
 	"syscall"
 
 	"github.com/systemboot/systemboot/pkg/storage"
 )
 
 // TODO backward compatibility for BIOS mode with partition type 0xee
-// TODO read and write non-volatile variables on the flash ROM via VPD
-//      https://chromium.googlesource.com/chromiumos/platform/vpd/
-//      via sysfs:
-//      https://github.com/torvalds/linux/blob/master/drivers/firmware/google/vpd.c
 // TODO use a proper parser for grub config (see grub.go)
 
 var (
-	baseMountPoint = flag.String("m", "/mnt", "Base mount point where to mount partiions")
+	baseMountPoint = flag.String("m", "/mnt", "Base mount point where to mount partitions")
 	doDryRun       = flag.Bool("dryrun", false, "Do not actually kexec into the boot config")
 	doDebug        = flag.Bool("d", false, "Print debug output")
+	doGrub         = flag.Bool("grub", false, "Use GRUB mode, i.e. look for valid Grub/Grub2 configuration in default locations to boot a kernel. GRUB mode ignores -kernel/-initramfs/-cmdline")
+	kernelPath     = flag.String("kernel", "", "Specify the path of the kernel to execute. If using -grub, this argument is ignored")
+	initramfsPath  = flag.String("initramfs", "", "Specify the path of the initramfs to load. If using -grub, this argument is ignored")
+	kernelCmdline  = flag.String("cmdline", "", "Specify the kernel command line. If using -grub, this argument is ignored")
+	deviceGUID     = flag.String("guid", "", "GUID of the device where the kernel (and optionally initramfs) are located. Ignored if -grub is set or if -kernel is not specified")
 )
 
 var debug = func(string, ...interface{}) {}
 
-func main() {
-	flag.Parse()
-
-	if *doDebug {
-		debug = log.Printf
-	}
-
-	// Get all the available block devices
-	devices, err := storage.GetBlockStats()
-	if err != nil {
-		log.Fatal(err)
-	}
-	// print partition info
-	for _, dev := range devices {
-		log.Printf("Device: %+v", dev)
-		table, err := storage.GetGPTTable(dev)
-		if err != nil {
-			continue
-		}
-		log.Printf("  Table: %+v", table)
-		for _, part := range table.Partitions {
-			log.Printf("    Partition: %+v", part)
-			if !part.IsEmpty() {
-				log.Printf("      UUID: %s", part.Type.String())
-			}
-		}
-	}
-
+// BootGrubMode tries to boot a kernel in GRUB mode. GRUB mode means:
+// * look for every attached storage device
+// * try to mount every device using any of the kernel-supported filesystems
+// * look for a GRUB configuration in various well-known locations
+// * build a list of valid boot configurations from the found GRUB configuration files
+// * try to boot every valid boot configuration until one succeeds
+//
+// The first parameter, `mountedDevices` is a list of storage.Mountpoint
+// representing a mounted storage device.
+// The second parameter, `dryrun`, will not boot the found configurations if set
+// to true.
+func BootGrubMode(devices []storage.BlockDev, baseMountpoint string, dryrun bool) error {
```

Comment:
baseMountpoint hides the outer var, not a huge issue, but maybe consider renaming it

Reply:
good point, renamed all the flags to be `flag<NAME>`